h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -240,69 +266,99 @@ protected MP getModelParams(MP params, Object[] hyperParams) {
     }
 
     protected long computeMaxSizeOfHyperSpace() {
-      long work = 1;
-      for (Map.Entry<String, Object[]> p : _hyperParams.entrySet()) {
-        if (p.getValue() != null) {
-          work *= p.getValue().length;
+      long work = 0;
+      long free_param_combos = 1;
+
+      for (Map<String, Object[]> subspace : _hyperParamSubspaces) {
+        long subspace_param_combos = 1;
+        for (Object[] o : subspace.values()) {
+          subspace_param_combos *= o.length;
         }
+        work += subspace_param_combos;
+      } // work will be zero if there is no subspaces in hyper parameters
+
+      for (Object[] p : _hyperParams.values()) {
+        free_param_combos *= p.length;
       }
+
+      work = work == 0 ? free_param_combos : free_param_combos * work;
+
       return work;
     }
 
+    protected Map<String, Object[]> mergeHashMaps(Map<String, Object[]> hyperparams, Map<String, Object[]> subspaces) {
```

Comment:
we defined:
```java
protected Map<String, Object[]>[] _hyperParamSubspaces;
```
so, please try to keep plural for array, and singular for elements of the array: here 
```java
Map<String, Object[]> subspace
```
small details, but when you read code on and on, it saves some brain processing :)